### ISSUE TITLE
learn_many vs learn_one

### ISSUE THREAD

**Issue by stirup** [online-ml/river#651]
## Versions

**river version**: 0.7.1
**Python version**: 3.8.10
**scikit-learn**:  0.24.2
**Operating system**: Redhat

## Describe the bug

Transformation and learning with `learn_one` and `learn_many` gives different output. May be I am missing something elementary in understanding how `learn_one` and `learn_many` works. 

## Steps/code to reproduce
```python
# Sample code to reproduce the problem
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.metrics import mean_squared_error
from river import stream
from river import preprocessing, linear_model, optim

print("Boston housing dataset prediction.")
boston = load_boston()
boston_data, boston_target = boston.data, boston.target

df_X = pd.DataFrame(boston_data, columns=boston.feature_names)
df_y = pd.Series(boston_target)


###### One
xtrain, xtest, ytrain, ytest = train_test_split(df_X, df_y, test_size=0.2, random_state=42)
stdScaler = preprocessing.StandardScaler()

model = linear_model.LinearRegression(
                 optimizer=optim.RMSProp())

dataset = stream.iter_pandas(xtrain, ytrain)

for x, yi in dataset:
    stdScaler.learn_one(x)
    x = stdScaler.transform_one(x)
    model.learn_one(x, yi)

ypred_one = []
dataset = stream.iter_pandas(xtest, ytest)

for x, yi in dataset:
    x = stdScaler.transform_one(x)
    ypred_one.append(model.predict_one(x))


mse = mean_squared_error(ytest, ypred_one)
print("MSE_one: ", mse)
print("RMSE_one: ", mse*(1/2.0))

##### Many

xtrain, xtest, ytrain, ytest = train_test_split(df_X, df_y, test_size=0.2, random_state=42)

stdScaler_many = preprocessing.StandardScaler()

model_many = linear_model.LinearRegression(
                 optimizer=optim.RMSProp())

dataset = stream.iter_pandas(xtrain, ytrain)

x = xtrain
yi = ytrain

stdScaler_many.learn_many(x)
x = stdScaler_many.transform_many(x)
model_many.learn_many(x, yi)

ypred_many = []
dataset = stream.iter_pandas(xtest, ytest)

x = xtest
x = stdScaler_many.transform_many(x)
ypred_many = model_many.predict_many(x)


mse = mean_squared_error(ytest, ypred_many)
print("MSE_many: ", mse)
print("RMSE_many: ", mse*(1/2.0))

x_ax = range(len(ytest))
plt.plot(x_ax, ytest, label="original")
plt.plot(x_ax, ypred_one, label="pred_one")
plt.plot(x_ax, ypred_many, label="pred_many")
plt.title("Boston test and predicted data")
plt.xlabel('X-axis')
plt.ylabel('Y-axis')
plt.legend(loc='best',fancybox=True, shadow=True)
plt.grid(True)
plt.show()
```

This is the output I get:
```
Boston housing dataset prediction.
MSE_one:  26.60962337529214
RMSE_one:  13.30481168764607
MSE_many:  505.2817051566025
RMSE_many:  252.64085257830126
```
Thank you.

**Comment by gbolmier** [comment]
Hi @stirup! It's natural that `learn_one` and `learn_many` lead to different set of weights. The gradient of a given data point is dependent of the current weights of the model. Therefore the gradients will be different if you compute them for multiple data points at the same time (against the same weights) or if you compute them one data point at a time i.e.: 

- compute the current data point's gradient
- **update the model's weights**
- repeat with the next data points

What you're doing with `learn_many` in the code sample you provided is batch gradient descent, because you're fitting the whole dataset at once. We can see that It performs poorly here with only one learning iteration.
